File: study/lua/src/app/ui/PopupActiveShopItemBuy.lua
--弹出界面
--通用活动商店购买
local PopupItemUse = require("app.ui.PopupItemUse")
local Path = require("app.utils.Path")
local TypeConvertHelper = require("app.utils.TypeConvertHelper")
local PopupActiveShopItemBuy = class("PopupActiveShopItemBuy", PopupItemUse)
local UserDataHelper = require("app.utils.UserDataHelper")
local ShopActiveDataHelper = require("app.utils.data.ShopActiveDataHelper")

function PopupActiveShopItemBuy:ctor(goodId, callback, title)
	self._title = title or Lang.get("shop_pop_title") 
	self._goodId = goodId
	self._callback = callback
	self._costResInfo1 = nil --消耗资源
	self._useNum  = 1 

	PopupActiveShopItemBuy.super.ctor(self, self._title, callback)
end

function PopupActiveShopItemBuy:onInitCsb()
	local CSHelper = require("yoka.utils.CSHelper")
	local resource = {
		file = Path.getCSB("PopupItemBuy", "common"),
		binding = {
			_btnOk = {
				events = {{event = "touch", method = "onBtnOk"}}
			},
			_btnCancel = {
				events = {{event = "touch", method = "onBtnCancel"}}
			},
		}
	}
   if resource then
        CSHelper.createResourceNode(self, resource)
    end
end


--
function PopupActiveShopItemBuy:onCreate()
	PopupActiveShopItemBuy.super.onCreate(self)

	self._costResInfo2:setVisible(false)
end


function PopupActiveShopItemBuy:onEnter()
    self:_updateUI()
    local maxLimit = ShopActiveDataHelper.getMaxLimit(self._goodId)
    self:setMaxLimit(maxLimit)
end

function PopupActiveShopItemBuy:onExit()
    
end

function PopupActiveShopItemBuy:_onNumSelect(num)
	self._useNum = num
	local info1 = self:_getItemPrice(1)
	if info1 then
		self:setCostInfo1(info1.type, info1.value, info1.size)
	end

    local info2 = self:_getItemPrice(2)
	if info2 then
		self:setCostInfo2(info2.type, info2.value, info2.size)
	end
end

function PopupActiveShopItemBuy:_getItemPrice(index)
	local costInfo = ShopActiveDataHelper.getCostInfo(self._goodId)
	local info = costInfo[index]
	if info then
		info.size = info.size * self._useNum
	end

	return info
end

function PopupActiveShopItemBuy:_updateUI()
	local info = ShopActiveDataHelper.getShopActiveConfig(self._goodId)
	local itemOwnerNum = UserDataHelper.getNumByTypeAndValue(info.type, info.value)
	PopupActiveShopItemBuy.super.updateUI(self, info.type, info.value, info.size)
	
	self:setOwnerCount(itemOwnerNum)
	self:_onNumSelect(self._useNum)
end

function PopupActiveShopItemBuy:onBtnOk()
	if self._callback then
		self._callback(self._goodId, self._useNum)
	end
	self:close()
end


function PopupActiveShopItemBuy:setCostInfo1(costType,costValue, costSize)
	self._costResInfo1:updateUI(costType,costValue,costSize)
end

function PopupActiveShopItemBuy:setCostInfo2(costType,costValue, costSize)
	self._costResInfo2:updateUI(costType,costValue,costSize)
	self._costResInfo2:setVisible(true)
end

return PopupActiveShopItemBuy

File: study/ts/script/ui/popup/PopupActiveShopItemBuy.ts
import PopupBase from "../PopupBase";
import PopupItemUse from "../PopupItemUse";
import { ShopActiveDataHelper } from "../../utils/data/ShopActiveDataHelper";
import { UserDataHelper } from "../../utils/data/UserDataHelper";
import { Lang } from "../../lang/Lang";
import CommonResourceInfo from "../component/CommonResourceInfo";
import { handler } from "../../utils/handler";
const { ccclass, property } = cc._decorator;

@ccclass
export default class PopupActiveShopItemBuy extends PopupItemUse{
    private _goodId:number;
    
    @property({ type: CommonResourceInfo, visible: true })
    _costResInfo1: CommonResourceInfo = null;
    @property({ type: CommonResourceInfo, visible: true })
    _costResInfo2: CommonResourceInfo = null;


    setInitData(goodId, callback, title?) {
        this.ctor(title,callback)
        this._goodId = goodId;
        this._useNum = 1; 
    }
    onInitCsb() {
        // var CSHelper = require('CSHelper');
        // var resource = {
        //     file: Path.getCSB('PopupItemBuy', 'common'),
        //     binding: {
        //         _btnOk: {
        //             events: [{
        //                     event: 'touch',
        //                     method: 'onBtnOk'
        //                 }]
        //         }
        //         _btnCancel: {
        //             events: [{
        //                     event: 'touch',
        //                     method: 'onBtnCancel'
        //                 }]
        //         }
        //     }
        // };
        // if (resource) {
        //     CSHelper.createResourceNode(this, resource);
        // }
    }
    onCreate() {
        super.onCreate();
        this._costResInfo2.setVisible(false);
    }
    onEnter() {
        super.onEnter()
        this._updateUI();
        var maxLimit = ShopActiveDataHelper.getMaxLimit(this._goodId);
        this.setMaxLimit(maxLimit);
    }
    onExit() {
    }
    _onNumSelect(num) {
        this._useNum = num;
        var info1 = this._getItemPrice(0);
        if (info1) {
            this.setCostInfo1(info1.type, info1.value, info1.size);
        }
        var info2 = this._getItemPrice(1);
        if (info2) {
            this.setCostInfo2(info2.type, info2.value, info2.size);
        }
    }
    _getItemPrice(index) {
        var costInfo = ShopActiveDataHelper.getCostInfo(this._goodId);
        var info = costInfo[index];
        if (info) {
            info.size = info.size * this._useNum;
        }
        return info;
    }
    _updateUI() {
        var info = ShopActiveDataHelper.getShopActiveConfig(this._goodId);
        var itemOwnerNum = UserDataHelper.getNumByTypeAndValue(info.type, info.value);
       this.updateUI(info.type, info.value, info.size);
        this.setOwnerCount(itemOwnerNum);
        this._onNumSelect(this._useNum);
    }
    onBtnOk() {
        if (this._callback) {
            this._callback(this._goodId, this._useNum);
        }
        this.close();
    }
    setCostInfo1(costType, costValue, costSize) {
        this._costResInfo1.updateUI(costType, costValue, costSize);
    }
    setCostInfo2(costType, costValue, costSize) {
        this._costResInfo2.updateUI(costType, costValue, costSize);
        this._costResInfo2.setVisible(true);
    }
}